### PR TITLE
Fix race in several Sockets tests

### LINKED ISSUE #5234: UDP tests fail with invalid checksum
```
System.Net.Sockets.Tests.SendReceive.SendToRecvFromAsync_Single_Datagram_UDP_IPv4 [FAIL]
11:27:58         Assert.Equal() Failure
11:27:58         Expected: 0
11:27:58         Actual:   1161193313
11:27:58         Stack Trace:
11:27:58            d:\j\workspace\dotnet_corefx\windows_nt_release_prtest\src\System.Net.Sockets\tests\FunctionalTests\SendReceive.cs(121,0): at System.Net.Sockets.Tests.SendReceive.SendToRecvFromAsync_Datagram_UDP(IPAddress leftAddress, IPAddress rightAddress)
```

http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/windows_nt_release_prtest/682/console

### PR DESCRIPTION
**Pull request by ericeil**
These tests were comparing against some checksum values without waiting until those values were actually generated/stored.  This change adds synchronization to ensure that the "sender" side of the test is done before validating the received data.  A nice side benefit of this is that any assertions and cleanup in the sender will also complete before we return from the test case, keeping it better isolated from other tests.

Fixes #5234 

@CIPop

### CODE REVIEW

**Comment by ericeil** [comment]
Windows build failures seem to be an infrastructure issue: http://dotnet-ci.cloudapp.net/job/dotnet_corefx/job/windows_nt_debug_prtest/2979/

**Comment by ericeil** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please
test Innerloop Windows_NT Release Build and Test please

File: src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs [hunk 1]
```diff
@@ -174,6 +175,7 @@ private static void SendToRecvFromAPM_Datagram_UDP(IPAddress leftAddress, IPAddr
                     if (sentDatagrams == DatagramsToSend)
                     {
                         right.Dispose();
+                        senderFinished.SetResult(true);
                         return;
                     }
```
Comment (stephentoub) [inline src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs]: Should we wrap the body of this if with a try/catch that stores any exceptions (including from an assert) into the senderFinished tcs?  Then such errors would come out of the Wait call in the main body of the test.

**Comment by stephentoub** [comment]
One comment to consider a small improvement, otherwise LGTM.

File: src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs (hunk 1 above)
Comment (ericeil) [inline src/System.Net.Sockets.Legacy/tests/FunctionalTests/SendReceive.cs]: > wrap the body of this if with a try/catch 

Great idea!  Done.

**Comment by stephentoub** [comment]
LGTM. Thanks.

**Pull request merged**